### PR TITLE
feat: support min & max date, DD-MM-YYYY date format 

### PR DESCRIPTION
**Pull request by alaa-yahia**
Support the implementation of [LIBS-418](https://dhis2.atlassian.net/browse/LIBS-440) & [DHIS2-17520](https://dhis2.atlassian.net/browse/DHIS2-17520).

### Includes:
- Expose date validation from useDatePicker hook 
- Expose error & warning messages
- Change formatYyyyMmDd to support both `yyyy-mm-dd` & `dd-mm-yyyy`

[LIBS-418]: https://dhis2.atlassian.net/browse/LIBS-418?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[DHIS2-17520]: https://dhis2.atlassian.net/browse/DHIS2-17520?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/hooks/useDatePicker.spec.tsx [hunk 1]
```diff
@@ -620,3 +620,49 @@ it('should generate the correct calendar weeks when passed "Ethiopian" rather th
...
     ])
 })
+
+describe('validation rules', () => {
+    it('should  validate correct format', () => {
+        const onDateSelect = jest.fn()
+        const date = '2015-'
+        const options = {
+            // minDate: '2015-06-28',
+            // calendar: 'iso8601' as SupportedCalendar,
```
Comment (kabaros) [inline src/hooks/useDatePicker.spec.tsx:630]: remove commented out code please

File: src/hooks/useDatePicker.spec.tsx [hunk 2]
```diff
@@ -620,3 +620,49 @@ it('should generate the correct calendar weeks when passed "Ethiopian" rather th
...
+        const options = {}
+        const minDate = '2015-06-28'
+        const maxDate = '2018-06-28'
+        const renderedHook = renderHook(() =>
+            useDatePicker({ onDateSelect, date, options, minDate, maxDate })
+        )
+        const result = renderedHook.result?.current as UseDatePickerReturn
+
+        expect(result.isValid).toEqual(false)
+    })
```
Comment (kabaros) [inline src/hooks/useDatePicker.spec.tsx:667]: could you also add similar tests for ethiopic and nepali calendars please

File: src/types.ts [hunk 3]
```diff
@@ -16,3 +16,8 @@ export type ResolvedLocaleOptions = {
     numberingSystem: string
     weekDayFormat: WeekDayFormat
 }
+
+export type ValidationOptions = {
```
Comment (kabaros) [inline src/types.ts:20]: this type doesn't seem to be used anywhere, so can be removed

File: src/utils/date-string-regexp.ts [hunk 4]
```diff
@@ -1 +1,2 @@
-export const dateStringRegExp = /^(\d{4})([-./])(\d{2})(\2)(\d{2})$/
+export const dateStringRegExp =
```
Comment (kabaros) [inline src/utils/date-string-regexp.ts:1]: can you add a comment clarifying the regex (that it matches `dd-...` and `yy-...`)

File: src/utils/extract-date-parts-from-date-string.ts [hunk 5]
```diff
@@ -7,10 +7,25 @@ export function extractDatePartsFromDateString(dateString: string) {
         throw new Error(`Date string is invalid, received "${dateString}"`)
     }
 
-    const [, yearStr, , monthStr, , dayStr] = parts
+    let yearStr, monthStr, dayStr, format
+
+    if (parts[1]) {
```
Comment (kabaros) [inline src/utils/extract-date-parts-from-date-string.ts:12]: what does this check exactly? what is `parts[1]` in each case

File: src/utils/extract-date-parts-from-date-string.ts [hunk 6]
```diff
@@ -7,10 +7,25 @@ export function extractDatePartsFromDateString(dateString: string) {
...
+        // Match for YYYY-MM-DD
+        yearStr = parts[1]
+        monthStr = parts[3]
+        dayStr = parts[5]
+        format = 'YYYY-MM-DD'
+    } else {
```
Comment (kabaros) [inline src/utils/extract-date-parts-from-date-string.ts:18]: can we also have a test (from the hook) to validate the effect of passing a completely different format (something like `2222-2222-2222` for example.

I see you have logic to fall back to today's date .. so you can have something like [this](https://github.com/dhis2/multi-calendar-dates/blob/88c6dc94b9b2b1c0f458db1a6093c9180954e944/src/hooks/useDatePicker.spec.tsx#L11) to ensure the test is not dependent on the actual date

File: src/utils/validate-date-string.spec.ts [hunk 7]
```diff
@@ -1,146 +1,180 @@
...
+
+    it('should return an error message when year is out of supported range', () => {
+        expect(
+            validateDateString('2101.04.33', { calendar: 'nepali' })
+                .errorMessage
+        ).toBe('Year 2101 is out of range.')
     })
 })
+
+// @todo: add tests for warning scenarios
```
Comment (kabaros) [inline src/utils/validate-date-string.spec.ts:180]: add these tests please

**Review by kabaros — APPROVED** [review]
looks good to me .. pre-approving with some comments - the most important are to add tests for `nepali` and `ethiopic` since these follow a different path.

Since this is a PR to alpha, feel free to merge them (to test with the UI component) and handle them separately in a followup PR.

File: src/utils/extract-date-parts-from-date-string.ts (hunk 5 above)
Comment (alaa-yahia) [reply src/utils/extract-date-parts-from-date-string.ts:12]: @kabaros I initially wrote it like [this](https://github.com/dhis2/multi-calendar-dates/blob/114d9e3baf904ec60534d8718d8d5b74b29cd7b5/src/utils/extract-date-parts-from-date-string.ts), but it caused issues with Prettier, it adds `;` at the beginning of these lines.

`[, yearStr, , monthStr, , dayStr] = parts;`
 `[, , , , , , dayStr, , monthStr, , yearStr] = parts;`

Parts[1] equal to this part (\d{4}), which means the format provided starts with 4 numbers.

Comment (kabaros) [reply src/utils/extract-date-parts-from-date-string.ts:12]: ok .. let's have it like this for now. We can see if it can be refactored later.

**Pull request merged**